chunk_downloader.go

```diff
@@ -0,0 +1,593 @@
+	if resp.StatusCode != http.StatusOK {
+		b, err := ioutil.ReadAll(bufStream)
+		if err != nil {
+			return err
+		}
+		logger.Infof("HTTP: %v, URL: %v, Body: %v", resp.StatusCode, scd.ChunkMetas[idx].URL, b)
+		logger.Infof("Header: %v", resp.Header)
+		return &SnowflakeError{
+			Number:      ErrFailedToGetChunk,
+			SQLState:    SQLStateConnectionFailure,
+			Message:     errMsgFailedToGetChunk,
+			MessageArgs: []interface{}{idx},
+		}
+	}
+	return decodeChunk(scd, idx, bufStream)
+}
+
+func decodeChunk(scd *snowflakeChunkDownloader, idx int, bufStream *bufio.Reader) (err error) {
+	gzipMagic, err := bufStream.Peek(2)
+	if err != nil {
+		return err
+	}
+	start := time.Now()
+	var source io.Reader
+	if gzipMagic[0] == 0x1f && gzipMagic[1] == 0x8b {
+		// detects and uncompresses Gzip format data
+		bufStream0, err := gzip.NewReader(bufStream)
+		if err != nil {
+			return err
+		}
+		defer bufStream0.Close()
+		source = bufStream0
+	} else {
+		source = bufStream
+	}
+	st := &largeResultSetReader{
+		status: 0,
+		body:   source,
+	}
+	var respd []chunkRowType
+	if scd.getQueryResultFormat() != arrowFormat {
+		var decRespd [][]*string
+		if !CustomJSONDecoderEnabled {
+			dec := json.NewDecoder(st)
+			for {
+				if err := dec.Decode(&decRespd); err == io.EOF {
+					break
+				} else if err != nil {
+					return err
+				}
+			}
+		} else {
+			decRespd, err = decodeLargeChunk(st, scd.ChunkMetas[idx].RowCount, scd.CellCount)
+			if err != nil {
+				return err
+			}
+		}
+		respd = make([]chunkRowType, len(decRespd))
+		populateJSONRowSet(respd, decRespd)
+	} else {
+		ipcReader, err := ipc.NewReader(source)
+		if err != nil {
+			return err
+		}
+		arc := arrowResultChunk{
+			*ipcReader,
+			0,
+			int(scd.totalUncompressedSize()),
+			memory.NewGoAllocator(),
+		}
+		respd, err = arc.decodeArrowChunk(scd.RowSet.RowType)
+		if err != nil {
+			return err
+		}
+	}
+	logger.Debugf(
+		"decoded %d rows w/ %d bytes in %s (chunk %v)",
+		scd.ChunkMetas[idx].RowCount,
+		scd.ChunkMetas[idx].UncompressedSize,
+		time.Since(start), idx+1,
+	)
+
+	scd.ChunksMutex.Lock()
+	defer scd.ChunksMutex.Unlock()
+	scd.Chunks[idx] = respd
+	return nil
+}
+
+func populateJSONRowSet(dst []chunkRowType, src [][]*string) {
+	// populate string rowset from src to dst's chunkRowType struct's RowSet field
+	for i, row := range src {
+		dst[i].RowSet = row
+	}
+}
+
+type streamChunkDownloader struct {
+	fetcher        streamChunkFetcher
+	readErr        error
+	rowStream      chan []*string
+	Total          int64
+	ChunkMetas     []execResponseChunk
+	NextDownloader chunkDownloader
+	RowSet         rowSetType
+}
+
+func (scd *streamChunkDownloader) totalUncompressedSize() (acc int64) {
+	return -1
+}
+
+func (scd *streamChunkDownloader) hasNextResultSet() bool {
+	return scd.readErr == nil
+}
+
+func (scd *streamChunkDownloader) nextResultSet() error {
+	return scd.readErr
+}
+
+func (scd *streamChunkDownloader) start() error {
```

Comment:
This function is missing a lot of the timing logs we have in the previous implementation. We rely on those logs to track timings and performance. Can they be brought back so we can hook our logger into them?

Reply:
@sfc-gh-cshi FYI this was a hard requirement for the structured logging integration. We need to be able to drop timing logs into the downloader.